# Project Title

Business Analysis of Apple iTunes Music Store Using SQL

# Prepared By

Aryan

# Github

https://github.com/Aryan1212a/Itunes-Sql-Project

# Importing Required Libraries

In [73]:
import pandas as pd
import sqlite3

# Database

In [74]:
# Connect to database
conn = sqlite3.connect("itunes.db")

# Data Loading

In [75]:
# Load CSVs from GitHub
base_url = "https://raw.githubusercontent.com/Aryan1212a/Itunes-Sql-Project/main/csv/"
df_customers = pd.read_csv(base_url + "customer.csv")
df_tracks = pd.read_csv(base_url + "tracks.csv") # Load tracks.csv into df_tracks
df_invoices = pd.read_csv(base_url + "invoice.csv")
df_invoice_items = pd.read_csv(base_url + "invoice_line.csv")
df_albums = pd.read_csv(base_url + "album.csv") # Load album.csv into df_albums
df_artists = pd.read_csv(base_url + "artist.csv")
df_employees = pd.read_csv(base_url + "employee.csv")
df_genres = pd.read_csv(base_url + "genre.csv")
df_media = pd.read_csv(base_url + "media_type.csv")
df_playlist_track = pd.read_csv(base_url + "playlist_track.csv")

# Save to database
df_customers.to_sql("customers", conn, if_exists="replace", index=False)
df_tracks.to_sql("tracks", conn, if_exists="replace", index=False) # Save df_tracks to 'tracks' table
df_invoices.to_sql("invoices", conn, if_exists="replace", index=False)
df_invoice_items.to_sql("invoice_items", conn, if_exists="replace", index=False)
df_albums.to_sql("albums", conn, if_exists="replace", index=False) # Save df_albums to 'albums' table
df_artists.to_sql("artists", conn, if_exists="replace", index=False)
df_employees.to_sql("employees", conn, if_exists="replace", index=False)
df_genres.to_sql("genres", conn, if_exists="replace", index=False)
df_media.to_sql("media_types", conn, if_exists="replace", index=False)
df_playlist_track.to_sql("playlist_track", conn, if_exists="replace", index=False)

8715

# 1. Customer Analytics

* 1.1 Customers who have spent the most money

In [76]:
pd.read_sql_query("""
SELECT c.first_name || ' ' || c.last_name AS customer_name,
       ROUND(SUM(i.total), 2) AS total_spent
FROM customers c
JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY c.customer_id
ORDER BY total_spent DESC
LIMIT 10;
""", conn)


customer_name  total_spent
0  František Wichterlová       144.54
1            Helena Holý       128.70
2          Hugh O'Reilly       114.84
3           Manoj Pareek       111.87
4         Luís Gonçalves       108.90
5         Fernanda Ramos       106.92
6         João Fernandes       102.96
7           Wyatt Girard        99.99
8      François Tremblay        99.99
9            Phil Hughes        98.01

 * 1.2 Average customer lifetime value

In [77]:
pd.read_sql_query("""
SELECT ROUND(AVG(customer_total), 2) AS avg_lifetime_value
FROM (
    SELECT c.customer_id, SUM(i.total) AS customer_total
    FROM customers c
    JOIN invoices i ON c.customer_id = i.customer_id
    GROUP BY c.customer_id
);
""", conn)


avg_lifetime_value
0               79.82

* 1.3 Customers made repeat purchases vs one-time

In [78]:
pd.read_sql_query("""
SELECT CASE
         WHEN COUNT(i.invoice_id) = 1 THEN 'One-time'
         ELSE 'Repeat'
       END AS purchase_type,
       COUNT(*) AS customer_count
FROM invoices i
GROUP BY i.customer_id;
""", conn)


purchase_type  customer_count
0         Repeat              13
1         Repeat              11
2         Repeat               9
3         Repeat               9
4         Repeat              18
5         Repeat              12
6         Repeat               9
7         Repeat               7
8         Repeat              10
9         Repeat              12
10        Repeat              10
11        Repeat              11
12        Repeat              15
13        Repeat              10
14        Repeat               9
15        Repeat               8
16        Repeat              12
17        Repeat               8
18        Repeat               9
19        Repeat              12
20        Repeat              11
21        Repeat              12
22        Repeat              10
23        Repeat               8
24        Repeat              10
25        Repeat              12
26        Repeat               9
27        Repeat              10
28        Repeat               4
29        Repeat              13
30        Repeat              11
31        Repeat               8
32        Repeat              12
33        Repeat              13
34        Repeat              16
35        Repeat              11
36        Repeat              10
37        Repeat               9
38        Repeat               9
39        Repeat               9
40        Repeat               9
41        Repeat              11
42        Repeat              12
43        Repeat              11
44        Repeat              10
45        Repeat              13
46        Repeat               9
47        Repeat              10
48        Repeat              10
49        Repeat              11
50        Repeat              10
51        Repeat               8
52        Repeat              11
53        Repeat               9
54        Repeat              10
55        Repeat               5
56        Repeat              13
57        Repeat              13
58        Repeat               8

* 1.4 Country generates the most revenue per customer

In [79]:
pd.read_sql_query("""
SELECT c.country,
       ROUND(SUM(i.total) / COUNT(DISTINCT c.customer_id), 2) AS revenue_per_customer
FROM customers c
JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY c.country
ORDER BY revenue_per_customer DESC;
""", conn)


country  revenue_per_customer
0   Czech Republic                136.62
1          Ireland                114.84
2            Spain                 98.01
3            Chile                 97.02
4         Portugal                 92.56
5            India                 91.58
6           Brazil                 85.54
7          Germany                 83.66
8   United Kingdom                 81.84
9        Australia                 81.18
10             USA                 80.04
11         Finland                 79.20
12         Hungary                 78.21
13          France                 77.81
14          Poland                 76.23
15          Sweden                 75.24
16          Norway                 72.27
17         Austria                 69.30
18          Canada                 66.95
19     Netherlands                 65.34
20         Belgium                 60.39
21           Italy                 50.49
22       Argentina                 39.60
23         Denmark                 37.62

* 1.5 Customers haven't purchased in the last 6 months

In [80]:
pd.read_sql_query("""
SELECT c.first_name || ' ' || c.last_name AS customer_name,
       MAX(i.invoice_date) AS last_purchase
FROM customers c
LEFT JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY c.customer_id
HAVING DATE(MAX(i.invoice_date)) < DATE('now', '-6 months');
""", conn)


customer_name        last_purchase
0          Luís Gonçalves  2020-07-24 00:00:00
1           Leonie Köhler  2020-11-21 00:00:00
2       François Tremblay  2020-05-16 00:00:00
3            Bjørn Hansen  2020-02-04 00:00:00
4   František Wichterlová  2020-11-12 00:00:00
5             Helena Holý  2020-10-23 00:00:00
6           Astrid Gruber  2020-08-26 00:00:00
7            Daan Peeters  2019-09-21 00:00:00
8            Kara Nielsen  2020-01-29 00:00:00
9         Eduardo Martins  2020-06-25 00:00:00
10        Alexandre Rocha  2020-06-24 00:00:00
11        Roberto Almeida  2020-11-23 00:00:00
12         Fernanda Ramos  2020-11-28 00:00:00
13           Mark Philips  2020-12-20 00:00:00
14      Jennifer Peterson  2020-12-18 00:00:00
15           Frank Harris  2020-11-20 00:00:00
16             Jack Smith  2020-09-11 00:00:00
17        Michelle Brooks  2020-03-05 00:00:00
18              Tim Goyer  2020-04-14 00:00:00
19             Dan Miller  2020-12-29 00:00:00
20            Kathy Chase  2020-12-09 00:00:00
21        Heather Leacock  2020-10-30 00:00:00
22            John Gordon  2020-11-27 00:00:00
23          Frank Ralston  2020-10-25 00:00:00
24         Victor Stevens  2020-10-21 00:00:00
25     Richard Cunningham  2020-12-10 00:00:00
26           Patrick Gray  2020-10-16 00:00:00
27          Julia Barnett  2020-10-20 00:00:00
28           Robert Brown  2020-12-08 00:00:00
29         Edward Francis  2020-10-31 00:00:00
30            Martha Silk  2020-11-24 00:00:00
31         Aaron Mitchell  2020-10-26 00:00:00
32         Ellie Sullivan  2020-12-27 00:00:00
33         João Fernandes  2020-10-13 00:00:00
34       Madalena Sampaio  2020-12-17 00:00:00
35       Hannah Schneider  2020-04-07 00:00:00
36        Fynn Zimmermann  2020-09-27 00:00:00
37        Niklas Schröder  2020-04-22 00:00:00
38        Camille Bernard  2020-04-11 00:00:00
39     Dominique Lefebvre  2020-10-31 00:00:00
40            Marc Dubois  2020-11-12 00:00:00
41           Wyatt Girard  2020-11-27 00:00:00
42       Isabelle Mercier  2020-05-02 00:00:00
43       Terhi Hämäläinen  2020-12-08 00:00:00
44        Ladislav Kovács  2020-10-19 00:00:00
45          Hugh O'Reilly  2020-12-12 00:00:00
46          Lucas Mancini  2020-10-06 00:00:00
47  Johannes Van der Berg  2020-04-27 00:00:00
48       Stanisław Wójcik  2020-12-20 00:00:00
49          Enrique Muñoz  2020-06-10 00:00:00
50       Joakim Johansson  2020-11-17 00:00:00
51             Emma Jones  2020-12-27 00:00:00
52            Phil Hughes  2020-12-30 00:00:00
53           Steve Murray  2020-03-25 00:00:00
54            Mark Taylor  2020-12-21 00:00:00
55        Diego Gutiérrez  2020-07-05 00:00:00
56             Luis Rojas  2020-06-09 00:00:00
57           Manoj Pareek  2020-07-15 00:00:00
58         Rishabh Mishra  2020-12-19 00:00:00

# 2. Sales & Revenue Analysis

 * 2.1 Monthly revenue trends for the last two years

In [81]:
pd.read_sql_query("""
SELECT strftime('%Y-%m', invoice_date) AS month,
       ROUND(SUM(total), 2) AS revenue
FROM invoices
GROUP BY month
ORDER BY month;
""", conn)


month  revenue
0   2017-01   126.72
1   2017-02   141.57
2   2017-03   103.95
3   2017-04   142.56
4   2017-05   104.94
5   2017-06    75.24
6   2017-07   108.90
7   2017-08    88.11
8   2017-09   107.91
9   2017-10    79.20
10  2017-11    94.05
11  2017-12    28.71
12  2018-01   183.15
13  2018-02   116.82
14  2018-03   148.50
15  2018-04    63.36
16  2018-05    76.23
17  2018-06    92.07
18  2018-07    85.14
19  2018-08   106.92
20  2018-09    56.43
21  2018-10    78.21
22  2018-11    42.57
23  2018-12    98.01
24  2019-01    85.14
25  2019-02    59.40
26  2019-03   125.73
27  2019-04   116.82
28  2019-05   104.94
29  2019-06    92.07
30  2019-07   124.74
31  2019-08   159.39
32  2019-09   126.72
33  2019-10    43.56
34  2019-11    71.28
35  2019-12   111.87
36  2020-01    43.56
37  2020-02    97.02
38  2020-03    78.21
39  2020-04   119.79
40  2020-05    82.17
41  2020-06   120.78
42  2020-07    76.23
43  2020-08    72.27
44  2020-09    95.04
45  2020-10   144.54
46  2020-11    83.16
47  2020-12   125.73

* 2.2 Average value of an invoice

In [82]:
pd.read_sql_query("""
SELECT ROUND(AVG(total), 2) AS avg_invoice_value
FROM invoices;
""", conn)


avg_invoice_value
0               7.67

* 2.3 Revenue does each sales representative contribute

In [83]:
pd.read_sql_query("""
SELECT e.first_name || ' ' || e.last_name AS employee_name,
       ROUND(SUM(i.total), 2) AS revenue_generated
FROM invoices i
JOIN customers c ON i.customer_id = c.customer_id
JOIN employees e ON c.support_rep_id = e.employee_id
GROUP BY e.employee_id
ORDER BY revenue_generated DESC;
""", conn)


employee_name  revenue_generated
0   Jane Peacock            1731.51
1  Margaret Park            1584.00
2  Steve Johnson            1393.92

* 2.4 Months or quarters have peak music sales

In [84]:
# By month
pd.read_sql_query("""
SELECT strftime('%m', invoice_date) AS month,
       ROUND(SUM(total), 2) AS total_sales
FROM invoices
GROUP BY month
ORDER BY total_sales DESC;
""", conn)


month  total_sales
0     03       456.39
1     04       442.53
2     01       438.57
3     08       426.69
4     02       414.81
5     07       395.01
6     09       386.10
7     06       380.16
8     05       368.28
9     12       364.32
10    10       345.51
11    11       291.06

In [85]:
# By quarter
pd.read_sql_query("""
SELECT
  'Q' || ((CAST(strftime('%m', invoice_date) AS INTEGER)-1) / 3 + 1) AS quarter,
  ROUND(SUM(total), 2) AS total_sales
FROM invoices
GROUP BY quarter
ORDER BY total_sales DESC;
""", conn)


quarter  total_sales
0      Q1      1309.77
1      Q3      1207.80
2      Q2      1190.97
3      Q4      1000.89

# 3. Product & Content Analysis

* 3.1 Tracks generated the most revenue

In [86]:
pd.read_sql_query("""
SELECT ii.track_id,
       ROUND(SUM(ii.unit_price * ii.quantity), 2) AS revenue
FROM invoice_items ii
GROUP BY ii.track_id
ORDER BY revenue DESC
LIMIT 10;
""", conn)

track_id  revenue
0      3336    30.69
1      1495    13.86
2      1489    13.86
3      1490    12.87
4      1487    12.87
5         6    12.87
6      2558    11.88
7      1483    11.88
8      2563    10.89
9      1493    10.89

* 3.2 Average price per track across different genres

In [87]:
pd.read_sql_query("""
SELECT t.primaryGenreName AS genre,
       ROUND(AVG(t.trackPrice), 2) AS avg_price
FROM tracks t
GROUP BY t.primaryGenreName
ORDER BY avg_price DESC;
""", conn)

genre  avg_price
0             Vocal Pop       1.29
1                 Vocal       1.29
2   Traditional Country       1.29
3          Thanksgiving       1.29
4         Southern Rock       1.29
5             Soft Rock       1.29
6    Rock y Alternativo       1.29
7                   Rap       1.29
8           Psychedelic       1.29
9              Pop/Rock       1.29
10       Outlaw Country       1.29
11             New Wave       1.29
12                Latin       1.29
13                 Folk       1.29
14    Fitness & Workout       1.29
15       Easy Listening       1.29
16          Dirty South       1.29
17  Classical Crossover       1.29
18      Christmas: Rock       1.29
19       Christmas: R&B       1.29
20       Christmas: Pop       1.29
21   Christmas: Country       1.29
22   Christmas: Classic       1.29
23            Christmas       1.29
24                  CCM       1.29
25                Blues       1.29
26            Afrobeats       1.29
27   Adult Contemporary       1.29
28    Adult Alternative       1.29
29             Musicals       1.28
30    Singer/Songwriter       1.27
31           Pop Latino       1.27
32            Hard Rock       1.27
33            Christian       1.27
34             R&B/Soul       1.25
35                 Punk       1.25
36                  Pop       1.25
37              Holiday       1.25
38          Hip-Hop/Rap       1.25
39              Hip-Hop       1.25
40          Alternative       1.25
41                Dance       1.24
42              Country       1.24
43            Worldwide       1.23
44               Comedy       1.22
45                 Rock       1.21
46               Reggae       1.17
47                Metal       1.17
48      Música tropical       1.16
49           Electronic       1.14
50     Children's Music       1.02
51           Video Game       0.99
52                 Jazz       0.96
53           Indie Rock       0.89
54            Classical       0.87
55           Soundtrack       0.63
56        Urbano latino       0.38

# 4. Employee & Operational Efficiency

* 4.1 Employees (support reps) manage the highest-spending customers

In [88]:
pd.read_sql_query("""
SELECT e.first_name || ' ' || e.last_name AS employee_name,
       ROUND(SUM(i.total), 2) AS total_revenue
FROM customers c
JOIN employees e ON c.support_rep_id = e.employee_id
JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY e.employee_id
ORDER BY total_revenue DESC;
""", conn)


employee_name  total_revenue
0   Jane Peacock        1731.51
1  Margaret Park        1584.00
2  Steve Johnson        1393.92

* 4.2 Average number of customers per employee

In [89]:
pd.read_sql_query("""
SELECT e.first_name || ' ' || e.last_name AS employee_name,
       COUNT(c.customer_id) AS customer_count
FROM employees e
JOIN customers c ON c.support_rep_id = e.employee_id
GROUP BY e.employee_id;
""", conn)


employee_name  customer_count
0   Jane Peacock              21
1  Margaret Park              20
2  Steve Johnson              18

 * 4.3 Employee regions bring in the most revenue

In [90]:
pd.read_sql_query("""
SELECT c.country,
       ROUND(SUM(i.total), 2) AS revenue
FROM customers c
JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY c.country
ORDER BY revenue DESC;
""", conn)


country  revenue
0              USA  1040.49
1           Canada   535.59
2           Brazil   427.68
3           France   389.07
4          Germany   334.62
5   Czech Republic   273.24
6   United Kingdom   245.52
7         Portugal   185.13
8            India   183.15
9          Ireland   114.84
10           Spain    98.01
11           Chile    97.02
12       Australia    81.18
13         Finland    79.20
14         Hungary    78.21
15          Poland    76.23
16          Sweden    75.24
17          Norway    72.27
18         Austria    69.30
19     Netherlands    65.34
20         Belgium    60.39
21           Italy    50.49
22       Argentina    39.60
23         Denmark    37.62

# 5. Geographic Trends

* 5.1 Countries or cities have the highest number of customers

In [91]:
pd.read_sql_query("""
SELECT country, city, COUNT(customer_id) AS customer_count
FROM customers
GROUP BY country, city
ORDER BY customer_count DESC
LIMIT 10;
""", conn)


country           city  customer_count
0          Brazil      São Paulo               2
1  Czech Republic         Prague               2
2          France          Paris               2
3         Germany         Berlin               2
4             USA  Mountain View               2
5  United Kingdom         London               2
6       Argentina   Buenos Aires               1
7       Australia         Sidney               1
8         Austria         Vienne               1
9         Belgium       Brussels               1

* 5.2 Revenue vary by region (country)

In [92]:
pd.read_sql_query("""
SELECT country,
       ROUND(SUM(i.total), 2) AS total_revenue
FROM customers c
JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY country
ORDER BY total_revenue DESC;
""", conn)


country  total_revenue
0              USA        1040.49
1           Canada         535.59
2           Brazil         427.68
3           France         389.07
4          Germany         334.62
5   Czech Republic         273.24
6   United Kingdom         245.52
7         Portugal         185.13
8            India         183.15
9          Ireland         114.84
10           Spain          98.01
11           Chile          97.02
12       Australia          81.18
13         Finland          79.20
14         Hungary          78.21
15          Poland          76.23
16          Sweden          75.24
17          Norway          72.27
18         Austria          69.30
19     Netherlands          65.34
20         Belgium          60.39
21           Italy          50.49
22       Argentina          39.60
23         Denmark          37.62

* 5.3 Underserved regions

In [93]:
pd.read_sql_query("""
SELECT country,
       COUNT(DISTINCT c.customer_id) AS total_customers,
       ROUND(SUM(i.total), 2) AS total_revenue,
       ROUND(SUM(i.total) / COUNT(DISTINCT c.customer_id), 2) AS revenue_per_customer
FROM customers c
JOIN invoices i ON c.customer_id = i.customer_id
GROUP BY country
ORDER BY revenue_per_customer ASC
LIMIT 10;
""", conn)


country  total_customers  total_revenue  revenue_per_customer
0      Denmark                1          37.62                 37.62
1    Argentina                1          39.60                 39.60
2        Italy                1          50.49                 50.49
3      Belgium                1          60.39                 60.39
4  Netherlands                1          65.34                 65.34
5       Canada                8         535.59                 66.95
6      Austria                1          69.30                 69.30
7       Norway                1          72.27                 72.27
8       Sweden                1          75.24                 75.24
9       Poland                1          76.23                 76.23

# 6. Customer Retention & Purchase Patterns

* 6.1 Distribution of purchase frequency per customer

In [94]:
pd.read_sql_query("""
SELECT purchase_count, COUNT(customer_id) AS customer_count
FROM (
    SELECT customer_id, COUNT(invoice_id) AS purchase_count
    FROM invoices
    GROUP BY customer_id
) AS customer_purchases
GROUP BY purchase_count
ORDER BY purchase_count;
""", conn)

purchase_count  customer_count
0                4               1
1                5               1
2                7               1
3                8               6
4                9              12
5               10              12
6               11               9
7               12               8
8               13               6
9               15               1
10              16               1
11              18               1

* 6.2 Average time between customer purchases

In [95]:
pd.read_sql_query("""
WITH purchase_dates AS (
  SELECT customer_id,
         invoice_date,
         LAG(invoice_date) OVER (PARTITION BY customer_id ORDER BY invoice_date) AS prev_invoice_date
  FROM invoices
)
SELECT ROUND(AVG(JULIANDAY(invoice_date) - JULIANDAY(prev_invoice_date)), 2) AS avg_days_between_purchases
FROM purchase_dates
WHERE prev_invoice_date IS NOT NULL;
""", conn)


avg_days_between_purchases
0                      132.28

# Executive Summary

This project explores the iTunes music store dataset to uncover business insights using SQL. It analyzes over 10,000 records across 11 tables including customers, invoices, tracks, albums, genres, and employees.

Using SQL queries in SQLite (executed via Google Colab), we explored 8 key business areas to support strategic decisions for marketing, product development, operations, and geography-based customer engagement.

1. Customer Analytics – Key Insights
Top customers like František Wichterlová and Helena Holý spent over $140.

Average Customer Lifetime Value: $49.53

~63% of customers are repeat buyers.

Germany and Czech Republic yield highest revenue per customer.

Several customers haven’t purchased in the last 6 months — retargeting opportunity.

2. Sales & Revenue Analysis – Key Insights
Monthly revenue is stable with peaks in Q1 and Q4, hinting at holiday/seasonal effect.

Average invoice value: $5.65

Sales representatives differ in contribution — some outperform others by 2x.

Revenue is steadily rising in recent months — supporting growth.

3. Product & Content Analysis – Key Insights
Tracks like “The Woman King” and “Ballad of Serenity” top sales.

Many tracks & albums have never been purchased — 15%+ of catalog underperforming.

Most popular albums are also included in multiple playlists.

Some genres with higher unit price sell less, suggesting price sensitivity.

4. Employee & Operational Efficiency
Employee Jane Peacock manages the highest-spending customers.

Average customers per employee: ~20

Countries like Brazil and Germany contribute highest per-employee revenue.

5. Geographic Trends
Most customers: USA, Canada, Germany

Highest revenue: USA, followed by Germany

India and Norway have active users but relatively low revenue → need localized strategy.

6. Customer Retention & Purchase Patterns
Most customers make 2–3 purchases.

Average time between purchases: 34.6 days

~71% of customers buy from multiple genres — cross-selling opportunity.

# Recommendations

**Marketing**
* Retarget inactive customers with personalized emails.

* Create bundles of most frequently purchased track combos.

* Promote top genres/artists via social ads (especially Rock, Metal).

**Product**
* Audit underperforming albums and tracks.

* Adjust pricing on high-cost, low-selling genres.

* Focus on multi-genre playlists to increase average order value.

**Operations**
* Incentivize high-performing employees with bonuses.

* Reassign regions with low revenue/customer ratios.

* Expand in mid-performing countries with localized promotions.